parser/ParseBlockProperties.hpp

```diff
@@ -0,0 +1,331 @@
+                                                              Property::kType);
+    item->values_->push_back(type_string);
+    return item;
+  }
+
+  /**
+   * @return Name of the parser construct represented by this class.
+   */
+  std::string getName() const override {
+    return std::string("BlockPropertyItem");
+  }
+
+  /**
+   * @brief The type of property being described.
+   *
+   * @return The specific block property which this instance is describing.
+   */
+  Property property() const {
+    return property_;
+  }
+
+  /**
+   * @return A list of ParseString values which the parser has captured to
+   *         describe this property. This may be empty in the case of 'COMPRESS ALL'.
+   */
+  const PtrList<ParseString>& values() const {
+    return *(values_.get());
+  }
+
+  /**
+   * @brief Gets the numeric value.
+   * @details Only set on the kBlockSize type.
+   *
+   * @return Pointer to the parsed numeric value.
+   */
+  const NumericParseLiteralValue* numeric_value() const {
+    return numeric_value_.get();
+  }
+
+  /**
+   * @brief Gets a string version of the property which this pair describes.
+   * @details For example, if this the property type of 'kCompress' will return
+   *          a string object with the value 'compress'.
+   *
+   * @return A string description of the property.
+   */
+  std::string getPropertyString() const {
+    switch (property_) {
+      case Property::kBlockSize:
+        return "blocksize";
+      case Property::kCompress:
+        return "compress";
+      case Property::kSort:
+        return "sort";
+      case Property::kType:
+        return "type";
+      default:
+        return "unknown";
+    }
+  }
+
+  /**
+   * @brief Returns if the user specified to compress all attributes (COMPRESS
+   *  ALL) attributes using a special keyword.
+   *
+   * @return \ctrue if this is a compression property and the value is 'ALL'.
+   */
+  bool compressAll() const {
+    return property_ == Property::kCompress && compress_all_;
+  }
+
+  /**
+   * @return True if it has parse string values.
+   */
+  bool hasStringValues() const {
+    return !hasNumericValue();
+  }
+
+  /**
+   * @return True if it has a numeric value.
+   */
+  bool hasNumericValue() const {
+    return property_ == Property::kBlockSize;
+  }
+
+ protected:
+  void getFieldStringItems(
+      std::vector<std::string> *inline_field_names,
+      std::vector<std::string> *inline_field_values,
+      std::vector<std::string> *non_container_child_field_names,
+      std::vector<const ParseTreeNode*> *non_container_child_fields,
+      std::vector<std::string> *container_child_field_names,
+      std::vector<std::vector<const ParseTreeNode*>> *container_child_fields) const override {
+    inline_field_names->push_back("property");
+    inline_field_values->push_back(getPropertyString());
+
+    if (hasStringValues()) {
+      container_child_field_names->push_back((values_->size() == 1) ? "value" : "values");
```

Comment:
It's probably easier (if not 100% grammatically correct from a human perspective) to always use the plural "values" (that way it's easier to write tools that parse the output of a serialized tree).